core/include/traccc/fitting/kalman_filter/gain_matrix_smoother.hpp

```diff
@@ -44,67 +44,117 @@ struct gain_matrix_smoother {
 
-        const auto& next_smoothed = next_state.smoothed();
-        const auto& next_predicted = next_state.predicted();
-        const auto& cur_filtered = cur_state.filtered();
-
-        // Next track state parameters
-        const matrix_type<6, 6>& next_jacobian = next_state.jacobian();
-        const matrix_type<6, 1>& next_smoothed_vec = next_smoothed.vector();
-        const matrix_type<6, 6>& next_smoothed_cov = next_smoothed.covariance();
-        const matrix_type<6, 1>& next_predicted_vec = next_predicted.vector();
-        const matrix_type<6, 6>& next_predicted_cov =
-            next_predicted.covariance();
-
-        // Current track state parameters
-        const matrix_type<6, 1>& cur_filtered_vec = cur_filtered.vector();
-        const matrix_type<6, 6>& cur_filtered_cov = cur_filtered.covariance();
-
-        // Regularization matrix for numerical stability
-        static constexpr scalar_type epsilon = 1e-13f;
-        const matrix_type<6, 6> regularization =
-            matrix_operator().template identity<e_bound_size, e_bound_size>() *
-            epsilon;
-        const matrix_type<6, 6> regularized_predicted_cov =
-            next_predicted_cov + regularization;
-
-        // Calculate smoothed parameter for current state
-        const matrix_type<6, 6> A =
-            cur_filtered_cov * matrix_operator().transpose(next_jacobian) *
-            matrix_operator().inverse(regularized_predicted_cov);
-
-        const matrix_type<6, 1> smt_vec =
-            cur_filtered_vec + A * (next_smoothed_vec - next_predicted_vec);
-        const matrix_type<6, 6> smt_cov =
-            cur_filtered_cov + A * (next_smoothed_cov - next_predicted_cov) *
-                                   matrix_operator().transpose(A);
-
-        cur_state.smoothed().set_vector(smt_vec);
-        cur_state.smoothed().set_covariance(smt_cov);
-
-        // projection matrix
-        // @TODO: Need to check if it is right to put the filtered state here
-        const typename mask_group_t::value_type::projection_matrix_type H =
-            mask_group[index].projection_matrix(cur_filtered);
-
-        constexpr const unsigned int D =
-            mask_group_t::value_type::shape::meas_dim;
-
-        // Calculate smoothed chi square
-        const matrix_type<D, 1>& meas_local = cur_state.measurement_local();
-        const matrix_type<D, D>& V = cur_state.measurement_covariance();
-        const matrix_type<D, 1> residual = meas_local - H * smt_vec;
-        const matrix_type<D, D> R =
-            V - H * smt_cov * matrix_operator().transpose(H);
-        const matrix_type<1, 1> chi2 = matrix_operator().transpose(residual) *
-                                       matrix_operator().inverse(R) * residual;
-
-        cur_state.smoothed_chi2() = matrix_operator().element(chi2, 0, 0);
+        this->operator()(cur_state, next_state);
     }
+
+    template <size_type D = 2u>
+    TRACCC_HOST_DEVICE inline void operator()(
```

Comment:
This maybe a stupid question, but where is ```D``` passed in the template parameter list when this operator is called in l. 51 and why does this need to be a separate operator from the call operator above?

Reply:
Fixed I guess